### DISCUSSION TITLE
Fan entity model speed lists

### DISCUSSION THREAD

**Discussion by finity69x2** [home-assistant/architecture#562] [Entity Models]
Is there any way we could incorporate a speed list back into the standard fan model for fans that have a limited number of speeds. as in the standard "low, medium, high" that three speed fans have been almost using since multi-speed electric fans have been created.

Using voice assistants saying "set the fan to 33%" isn't natural for the fans that most people grew up around.

Of course leave those percentage options along with the speeds as well.

Or even better would be to incorporate the percentages into the speed list.

For example you could use:

```
speeds:
  - 'low'
  - 'medium'
  - 'high'
```

or

```
speeds:
  - '25%'
  - '50%'
  - '75%'
  - '100%'
```

As it is now users have to derive their percentage steps from a math calculation after trying to figure out what the different code options mean.

**Comment by frenck** [comment]
@finity69x2 What is your implementation suggestion? As in, if both are available how would this be determined?
Could you sketch a general idea of the adjustment in code for this? Do you have a branch we can look at somewhere?

While external integrators may be weird in voice commands, their UI's (e.g., HomeKit) actually often do only work correctly with percentages. And it is not only voice we have do deal with, it is also all other interface and/or device capabilities.

**Reply by finity69x2** [reply]
I don't have any code implementation for this just a general idea.

we already have a preset_modes list so that would be kept as is.

a named_speeds list and a percent_speeds list could be added.

```
named_speeds:
  - low
  - medium
  - high
  - super
```

```
percent_speeds:
  - 10
  - 40
  - 70
  - 100
```

All of those would end up being added into the Alexa (in my case) speed_list attribute so that it could understand all of the options. Right now Alexa can't control my MQTT fans using preset or percentages because the fan doesn't contain a speed list. And The Alexa App can't even discover the fan entity when configured using either of those two configuration methods either.  I submitted an issue report [here](https://github.com/home-assistant/core/issues/50198) 

Then in the config the "speed" related configuration options ("speed_value_template", "speed_command_template", etc and for the MQTT fan the speed topics if needed) should be added back in (along with those related services) so that whatever implementation the user needs they have it available. 

It would be very similar to what we have now in that we can configure both preset_modes and percentages in the same fan config. So there could be preset_modes and named_speeds, preset_modes and percentage, or etc.

Of course, the current options for percentages could all still be left in place to allow users who need it to take advantage of their infinitely variable speed fans.

We already also still have a "sort of" similar functionality right now since my 3 speed mqtt fans still work using the old speed list configuration and with either of the preset mode or percentage speeds configuration. The only thing that is going on right now with the old speed config is that when the fan speed is switched via the fan.set_speed service there is a warning in the log that says that the service is deprecated and needs to be updated to the new config. I believe (without looking into the code of course) that the functionality already exists to work the way I think it could so it should take very little work to implement. Simply remove the deprecation warning for the service call and update the documentation to add the named speed related stuff back in as options.

And not to mention that right now I have a Zigbee fan that still has all of the named speeds along with the a couple of preset modes that still works with the Alexa voice assistant. So doing this would maintain consistency throughout the config and especially in the expectation that the users know how the fans are supposed to work across all platforms.

Thanks for considering this.

**Reply by frenck** [reply]
How would this work with e.g., HomeKit? As that uses a percentages slider.

I have to ask, are you planning on implementing this?

**Reply by finity69x2** [reply]
I believe all of the percentage functionalities would still be available as long as the new percentage configuration options are left intact. I don't have Homekit but I can't imagine the two models can't work side by side since they seem to be happily doing that right now since the old model hasn't yet been removed.

And TBH, I wouldn't know how to implement this myself. I don't have a problem trying but I'm pretty sure my skills aren't up to the task right now.

But I just looked at the code in the current HA fan integration and it seems that all of the pieces are in place right now to implement this since the old speed configs still work (but with a warning about the "set_speed" service deprecation) and the new percentage services work including the sliders. 

I can imagine that all that needs done is to undo the translation from "speed list" to "percentages" that is now happening in the background. And of course to stop the currently planned deprecation of the set_speed service along with removing the warning for the (hopefully no longer) deprecated set_speed service.

I get that you don't necessarily want the "non-speed" modes to be used as speeds but it seems there must be a way for voice assistants to understand a "normal" speed command (i.e. use named speeds) and not require using a "percent" speed and still allow the percent speeds to be used where a user wants to use them. I'm not suggesting at all to remove the percentages functionality. Just to make the speeds and percentages work in parallel as needed.

And again, as it is right now the Alexa integration is completely broken when trying to use the new fan models because the Alexa app is expecting a speed_list that isn't provided by the current fan integration. So either way work will need to be done to fix the Alexa integration but I think it will be less work to leave the speed_list intact and then figure out how to incorporate percentages and preset_modes so Alexa understands those as well.

**Reply by frenck** [reply]
> And TBH, I wouldn't know how to implement this myself.

Right, but this is the architectural repository in case you want to implement a change and need permission for that.
If you want to do feature requests, please use the community forums instead.

**Reply by finity69x2** [reply]
Sorry, I was told in the issue I submitted that any discussions of the kind for one statement I made should be made here and they gave me the link. 

I guess I didn't understand what this was for. I thought it was for questions about architectural decisions and suggestions related to them.

Especially since I was sent here by a developer there.

**Reply by frenck** [reply]
This is not the community forum. If you want changes you cannot implement yourself, then use the feature request section.

If you are creating/building changes to Home Assistant core, like changes in entity components like you are suggesting here, this can be discussed and approved here.

**Reply by thecode** [reply]
I think that it is better to support both speeds in percentage and a named list.
@finity69x2 please open a feature request here:
https://community.home-assistant.io/c/feature-requests/13
If there is high request for it I will get the discussion back here how to implement

**Reply by emontnemery** [reply]
We very recently moved from a (fixed, low, medium, high) list of named speeds to a speed in percentage **+ a `speed_count`**.

One key reason for the move was that the voice assistants and other external APIs all have different limitations, which are not perfectly aligned with all of the smart fans and fan controllers on the market.

A quick review of the voice assistant APIs as of today:
- Apple Homekit [supports only speed as percentage, no named speeds from what I understand](https://developer.apple.com/documentation/homekit/hmcharacteristictyperotationspeed)
- Amazon Alexa [supports a fixed number of speeds, with optional aliases for the speeds](https://developer.amazon.com/en-US/docs/alexa/device-apis/alexa-rangecontroller.html)
- Google Assistant [supports both speed as percentage, and a fixed number of named speeds](https://developers.google.com/assistant/smarthome/traits/fanspeed)

It could make sense to allow an integration to supply an ordered list of aliases for the speeds it supports, but we then have the problem of translating the listed speed aliases to the language used by the voice assistant.

Another option is to let Alexa and Google Assistant create aliases, so something like:
- If the fan supports 2 speeds: low, high
- If the fan supports 3 speeds: low, medium, high
- If the fan supports 4-10 speeds: numbers 1..`speed_count`
- If the fan supports >10 speeds: no aliases, percentage only

**Reply by jbouwh** [reply]
Since we introduced `speeds_count` and ha a `percentage_step` attribute exposed, we can have as many speeds as needed, without a barrier.
The main task is to translate speeds like ZIGBEE MQTT speeds back to percentage and the preset_modes to presets.
The language is still a barrier in integrations as Google_assistant and Alexa. To make presets possible I made 2 PR's to introduce support for preset modes.
### Alexa
Alexa has a [Modes controller](https://developer.amazon.com/en-US/docs/alexa/device-apis/alexa-modecontroller.html) I used to implement fan `preset_mode`:
https://github.com/home-assistant/home-assistant.io/pull/17800

As language only US English is supported right now.

### Google Assistant
The Google Assistant API has a [Modes trait](https://developers.google.com/assistant/smarthome/traits/modes) that I used to the fan `preset_mode`:
https://github.com/home-assistant/core/pull/50283

Google an Alexa can translate speeds from the natural language to a percentage out of the box. So from that point of view there is not a big need to introduce named speeds. Preset modes can help to put a fan into automatic or smart operation.

Now the classic speeds have been deprecated (and we do not want those back) it would be very confusing introducing them back again. To my opinion percentage based speeds with preset_modes will work for most use cases.

Before we introduce named attributes, a good solution for handling locales and synonyms is needed .

**Reply by emontnemery** [reply]
Right, what I had mind was:
- The fan MUST support getting and setting speed based on a percentage (no change)
- The fan MUST export a speed_count property (no change)
- The fan can OPTIONALLY export a sorted `speed_list` or `speed_names` of length `speed_count` with synonyms for the speeds, with the purpose of teaching [alexa's rangecontroller](https://developer.amazon.com/en-US/docs/alexa/device-apis/alexa-rangecontroller.html) and [google's fanspeed](https://developers.google.com/assistant/smarthome/traits/fanspeed) about the named speeds

**Reply by bdraco** [reply]
> Right, what I had mind was:
> 

> - The fan can OPTIONALLY export a sorted `speed_list` or `speed_names` of length `speed_count` with synonyms for the speeds, with the purpose of teaching [alexa's rangecontroller](https://developer.amazon.com/en-US/docs/alexa/device-apis/alexa-rangecontroller.html) and [google's fanspeed](https://developers.google.com/assistant/smarthome/traits/fanspeed) about the named speeds

There are a few things that came up in the last round of discussion where similar ideas were presented:

- MUST: make them translatable (unless they are user configured from yaml etc)
- SHOULD: add validation to prevent invalid mapping sizes
- NICE TO HAVE: come up with a way to detect mixing preset modes in the list of names as this was frequently a problem in the old model

I agree that whatever we come up with, it should be OPTIONAL as we still have a lot of fans sitting in HACS because the old model was restrictive to 3 speeds and we don't want to increase the barrier any more to get them into core.
